src/lightning/fabric/utilities/throughput_monitor.py

```diff
@@ -0,0 +1,354 @@
+        flops_per_batch: Optional[int] = None,  # (per device)
+        lengths: Optional[int] = None,  # total length of the samples seen (per device)
+    ) -> None:
+        self._step += 1
+        step = self._step
+        metrics = {}
+
+        self._samples.append(samples)
+        if lengths is not None:
+            self._lengths.append(lengths)
+            # if lengths are passed, there should be as many values as samples
+            assert len(self._samples) == len(self._lengths)
+        self._time.append(train_elapsed)
+        if len(self._time) == self._time.maxlen:
+            elapsed_batches = len(self._samples) - 1
+            elapsed_samples = self._samples[-1] - self._samples[0]
+            elapsed_wct = self._time[-1] - self._time[0]
+            samples_per_sec = elapsed_samples * world_size / elapsed_wct
+            dev_samples_per_sec = elapsed_samples / elapsed_wct
+            metrics.update(
+                {
+                    "batches_per_sec": elapsed_batches * world_size / elapsed_wct,
+                    "samples_per_sec": samples_per_sec,
+                    "device/batches_per_sec": elapsed_batches / elapsed_wct,
+                    "device/samples_per_sec": dev_samples_per_sec,
+                }
+            )
+            if lengths is not None:
+                elapsed_lengths = int(self._lengths[-1]) - int(self._lengths[0])
+                avg_length = elapsed_lengths / elapsed_batches
+                metrics.update(
+                    {
+                        "items_per_sec": samples_per_sec * avg_length,
+                        "device/items_per_sec": dev_samples_per_sec * avg_length,
+                    }
+                )
+
+        if flops_per_batch is not None:
+            # sum of flops per batch across ranks
+            self._flops.append(flops_per_batch * world_size)
+        if len(self._flops) == self._flops.maxlen:
+            elapsed_flops = sum(self._flops) - self._flops[0]
+            elapsed_wct = self._time[-1] - self._time[0]
+            flops_per_sec = elapsed_flops / elapsed_wct
+            device_flops_per_sec = flops_per_sec / world_size
+            metrics.update({"flops_per_sec": flops_per_sec, "device/flops_per_sec": device_flops_per_sec})
+            if self.flops_available:
+                metrics["device/mfu"] = device_flops_per_sec / self.flops_available
+
+        metrics.update(
+            {
+                "time/train": train_elapsed / self._divider,
+                "time/val": self._total_eval_time / self._divider,
+                "time/total": (train_elapsed + self._total_eval_time) / self._divider,
+                "samples": samples,
+            }
+        )
+
+        for logger in self.loggers:
+            logger.log_metrics(metrics=metrics, step=step)
+
+    def eval_end(self, eval_elapsed: float) -> None:
+        self._total_eval_time += eval_elapsed  # seconds
+
+
+class ThroughputMonitor(_ThroughputMonitorBase):
```

Comment:
I personally wish that I could use this as a utility by itself, without passing a logger into it and get the metrics dict back.

Reply:
yes, that would be nice

Reply:
You commented it on the `ThroughputMonitor` used by Fabric. Did you mean to make this suggestion on what is currently called `_ThroughtputMonitorBase`?

If so, the logger logic could be moved into the `ThroughputMonitor` in both PL and Fabric. What would you name `_ThroughtputMonitorBase` then?

Reply:
Yes I meant to comment it here on the Fabric version, where I would prefer to interact with it directly to get the metrics. It's great with the latest change to return the metric dict!

Yes, I agree the PL version would do the logging automatically though. Whether there should be a common base class, I am not sure.

One other way could be that the `ThroughputMonitor` from Fabric is used with composition inside the PL callback, where the callback only does book keeping and logging, but calls into the monitor from fabric.

Reply:
This is done now.

`ThroughputMonitor` in PL is a regular callback.
`_ThroughputMonitor` in Fabric is a utility class (name suggestions welcome if we want to make it public)
`ThroughputMonitor` in Fabric is a utility class that integrates `_ThroguhputMonitor` with `Fabric` and manages logging too.